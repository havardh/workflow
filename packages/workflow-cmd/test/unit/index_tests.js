/* eslint-env node, jest */
import execa from 'execa';
import { join } from 'path';

const cli = join(__dirname, '..', '..', 'index.js');
const config = join(__dirname, '__fixtures__', 'config.js');

const n = n => Array(n + 1).join('-');
const format = (tag, out) => (out ? `${tag} ${n(73)}\n${out}\n ${n(80)}\n` : '');

async function run(command) {
  const args = command.replace(/^workflow /, '').match(/(?:[^\s"]+|"[^"]*")+/g);

  const { stdout, stderr } = await execa('node', [cli, ...args]);

  return format('stdout', stdout) + format('stderr', stderr);
}

describe('help', () => {
  it('should have flag: --help', async () => {
    const output = await run(`workflow  --config ${config} --help`);

    expect(output).toMatchSnapshot();
  });

  it('should have a command: help', async () => {
    const output = await run(`workflow  --config ${config} help`);

    expect(output).toMatchSnapshot();
  });

  it('should be available for the update command', async () => {
    const output = await run(`workflow --config ${config} update --help`);

    expect(output).toMatchSnapshot();
  });

  it('should be available for the apply command', async () => {
    const output = await run(`workflow --config ${config} apply --help`);

    expect(output).toMatchSnapshot();
  });

  it('should provide help for explicit flows with string argument', async () => {
    const config = join(__dirname, '__fixtures__', 'config.js');
    const flow = JSON.stringify({
      type: 'workflow',
      args: 'file',
      children: [],
    });

    const output = await run(`workflow --config ${config} apply ${flow} --help`);
    expect(output).toMatchSnapshot();
  });

  it('should provide help for flows with string argument', async () => {
    const flow = JSON.stringify({
      type: 'workflow',
      args: 'file',
      children: [],
    });

    const output = await run(`workflow --config ${config} ${flow} --help`);
    expect(output).toMatchSnapshot();
  });

  it('should provide help for flows with array argument', async () => {
    const flow = JSON.stringify({
      type: 'workflow',
      args: ['cwd', 'cmd'],
      children: [],
    });

    const output = await run(`workflow --config ${config} ${flow} --help`);
    expect(output).toMatchSnapshot();
  });

  it('should provide help for flows with object argument', async () => {
    const flow = JSON.stringify({
      type: 'workflow',
      args: { cwd: 'the current working dir', cmd: 'the command to execute' },
      children: [],
    });

    const output = await run(`workflow --config ${config} ${flow} --help`);
    expect(output).toMatchSnapshot();
  });
});

it('should render a flow', async () => {
  const flow = JSON.stringify({
    type: 'workflow',
    children: [],
  });
  const output = await run(`workflow --config ${config} ${flow}`);

  expect(output).toMatchSnapshot();
});
