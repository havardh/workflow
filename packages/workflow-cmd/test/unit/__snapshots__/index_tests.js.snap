// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`help should be available for the apply command 1`] = `
"stdout -------------------------------------------------------------------------
workflow apply <flow>

apply the flow

Positionals:
  flow  the flow file to load                                [string] [required]

Options:
  --version        Show version number                                 [boolean]
  --config, -c     Set the workflow-home by path to config.js           [string]
  --verbose, --vv  Set output mode to verbose                          [boolean]
  --help           Show help                                           [boolean]
 --------------------------------------------------------------------------------
"
`;

exports[`help should be available for the update command 1`] = `
"stdout -------------------------------------------------------------------------
workflow update

Update workflow and workflow-home dependencies

Options:
  --version        Show version number                                 [boolean]
  --config, -c     Set the workflow-home by path to config.js           [string]
  --verbose, --vv  Set output mode to verbose                          [boolean]
  --help           Show help                                           [boolean]
  --latest         Updates dependencies in workflow-home to @latest    [boolean]
  --force          Ignore uncommited check when updating workflow-home [boolean]
 --------------------------------------------------------------------------------
"
`;

exports[`help should have a command: help 1`] = `
"stdout -------------------------------------------------------------------------
workflow [<cmd>]

Commands:
  workflow version       Show the version
  workflow update        Update workflow and workflow-home dependencies
  workflow apply <flow>  apply the flow                                [default]

Positionals:
  flow  the flow file to load                                [string] [required]

Options:
  --version        Show version number                                 [boolean]
  --config, -c     Set the workflow-home by path to config.js           [string]
  --verbose, --vv  Set output mode to verbose                          [boolean]
  --help           Show help                                           [boolean]
 --------------------------------------------------------------------------------
"
`;

exports[`help should have flag: --help 1`] = `
"stdout -------------------------------------------------------------------------
workflow [<cmd>]

Commands:
  workflow version       Show the version
  workflow update        Update workflow and workflow-home dependencies
  workflow apply <flow>  apply the flow                                [default]

Positionals:
  flow  the flow file to load                                [string] [required]

Options:
  --version        Show version number                                 [boolean]
  --config, -c     Set the workflow-home by path to config.js           [string]
  --verbose, --vv  Set output mode to verbose                          [boolean]
  --help           Show help                                           [boolean]
 --------------------------------------------------------------------------------
"
`;

exports[`help should provide help for explicit flows with string argument 1`] = `
"stdout -------------------------------------------------------------------------
workflow apply <flow> <file>

apply the flow

Positionals:
  flow  the flow file to load                                [string] [required]
  file  file                                                 [string] [required]

Options:
  --version        Show version number                                 [boolean]
  --config, -c     Set the workflow-home by path to config.js           [string]
  --verbose, --vv  Set output mode to verbose                          [boolean]
  --help           Show help                                           [boolean]
 --------------------------------------------------------------------------------
"
`;

exports[`help should provide help for flows with array argument 1`] = `
"stdout -------------------------------------------------------------------------
workflow [<cmd>]

Commands:
  workflow version                   Show the version
  workflow update                    Update workflow and workflow-home
                                     dependencies
  workflow apply <flow> <cwd> <cmd>  apply the flow                    [default]

Positionals:
  flow  the flow file to load                                [string] [required]
  cwd   cwd                                                  [string] [required]
  cmd   cmd                                                  [string] [required]

Options:
  --version        Show version number                                 [boolean]
  --config, -c     Set the workflow-home by path to config.js           [string]
  --verbose, --vv  Set output mode to verbose                          [boolean]
  --help           Show help                                           [boolean]
 --------------------------------------------------------------------------------
"
`;

exports[`help should provide help for flows with object argument 1`] = `
"stdout -------------------------------------------------------------------------
workflow [<cmd>]

Commands:
  workflow version       Show the version
  workflow update        Update workflow and workflow-home dependencies
  workflow apply <flow>  apply the flow                                [default]

Positionals:
  flow  the flow file to load                                [string] [required]

Options:
  --version        Show version number                                 [boolean]
  --config, -c     Set the workflow-home by path to config.js           [string]
  --verbose, --vv  Set output mode to verbose                          [boolean]
  --help           Show help                                           [boolean]
  --cwd            the current working dir                            [required]
  --cmd            the command to execute                             [required]
 --------------------------------------------------------------------------------
"
`;

exports[`help should provide help for flows with string argument 1`] = `
"stdout -------------------------------------------------------------------------
workflow [<cmd>]

Commands:
  workflow version              Show the version
  workflow update               Update workflow and workflow-home dependencies
  workflow apply <flow> <file>  apply the flow                         [default]

Positionals:
  flow  the flow file to load                                [string] [required]
  file  file                                                 [string] [required]

Options:
  --version        Show version number                                 [boolean]
  --config, -c     Set the workflow-home by path to config.js           [string]
  --verbose, --vv  Set output mode to verbose                          [boolean]
  --help           Show help                                           [boolean]
 --------------------------------------------------------------------------------
"
`;

exports[`should render a flow 1`] = `
"stdout -------------------------------------------------------------------------
{
  \\"type\\": \\"workflow\\",
  \\"children\\": []
}
 --------------------------------------------------------------------------------
"
`;
